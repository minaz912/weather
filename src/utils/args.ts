import flow from 'lodash/fp/flow'
import { CategorizedArg } from '../types'

export function categorize(arg: string): CategorizedArg {
  return !Number.isNaN(Number(arg))
    ? { queryType: 'ZIP_CODE', queryValue: arg }
    : { queryType: 'CITY_NAME', queryValue: arg }
}

function join(str: string[]): string {
  return str.join('')
}

function trim(str: string): string {
  return str.trim()
}

function split(str: string): string[] {
  return str.split(',')
}

export function getArgList(processArgs: string[]): string[] {
  return flow(
    join,
    trim,
    split
  )(processArgs)
}
