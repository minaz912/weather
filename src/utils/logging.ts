export function showError(city: string, err?: Error): void {
  console.error(
    `City/ZipCode: ${city}, Error encountered: ${err || 'Unknown Error'}`
  )
}

export function showResult(
  city: string,
  time: string,
  temperature: string
): void {
  console.log(`City/ZipCode: ${city}, Time: ${time}, Weather: ${temperature}`)
}
