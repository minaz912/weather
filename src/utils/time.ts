import moment from 'moment-timezone';

export function getPrettyDateWithTzOffset(utcOffsetInSeconds?: number): string {
  return utcOffsetInSeconds
    ? moment()
        .utcOffset(utcOffsetInSeconds / 60)
        .format('LLL')
    : 'N/A';
}
