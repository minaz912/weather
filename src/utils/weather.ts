import map from 'lodash/fp/map'
import flow from 'lodash/fp/flow'
import axios, { AxiosPromise, AxiosResponse, AxiosError } from 'axios'

import * as args from './args'

import { CategorizedArg, RequestMetadata, WeatherResult } from '../types'

class OpenWeatherMapAPIManager {
  static fetcher = axios.create({
    // ideally here we would separate the api version from the base url
    baseURL: 'https://api.openweathermap.org/data/2.5',
    params: {
      // TODO: handle missing env variable
      APPID: process.env.OWM_APPID,
      units: 'metric'
    }
  })
}

function getRequestMetadata(queryMetadata: CategorizedArg): RequestMetadata {
  const { queryType, queryValue } = queryMetadata
  if (queryType === 'ZIP_CODE') {
    return {
      queryValue,
      endpoint: '/weather',
      params: {
        // OWM expects the zip code to be passed in a 'zip' query arg
        // ref: https://openweathermap.org/current#zip
        zip: queryValue
      }
    }
  }
  if (queryType === 'CITY_NAME') {
    return {
      queryValue,
      endpoint: '/weather',
      params: {
        // OWM expects the city name to be passed in a 'q' query arg
        // ref: https://openweathermap.org/current#name
        q: queryValue
      }
    }
  }

  throw new Error('Invalid metadata')
}

function getWeatherFromApi(requestMetadata: RequestMetadata): AxiosPromise {
  const { endpoint, params } = requestMetadata
  return OpenWeatherMapAPIManager.fetcher.get(endpoint, {
    params
  })
}

function generateWeatherMessage(
  apiResult: AxiosResponse | AxiosError,
  requestMetadata: RequestMetadata
): WeatherResult {
  if (apiResult instanceof Error) {
    return {
      name: requestMetadata.queryValue,
      weather: apiResult.response.data.message,
      timezoneOffsetInSeconds: null
    }
  }

  const {
    data: { name, weather, main, timezone }
  } = apiResult
  return {
    name,
    weather: `${weather[0].main}, ${main.temp} Celsius`,
    timezoneOffsetInSeconds: timezone
  }
}

export async function getWeatherResult(
  cities: string[]
): Promise<WeatherResult[]> {
  // try {
  const categorizedRequestMetadata = map((arg: string) =>
    flow(
      args.categorizeCityArg,
      getRequestMetadata
    )(arg)
  )(cities)

  return Promise.all(
    categorizedRequestMetadata.map(metadata =>
      getWeatherFromApi(metadata)
        .then(response => generateWeatherMessage(response, metadata))
        .catch(err => generateWeatherMessage(err, metadata))
    )
  )
}
