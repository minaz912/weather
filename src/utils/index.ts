import * as args from './args'
import * as logging from './logging'
import * as weather from './weather'
import * as time from './time'

export default {
  args,
  logging,
  time,
  weather
}
