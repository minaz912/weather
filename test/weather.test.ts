import mockAxios from 'axios'
import * as weather from '../src/utils/weather'

describe('getWeatherResult', () => {
  it('returns weather result messages based on given cities', async () => {
    const cities = ['Cairo', '10005', 'New York']

    const results = await weather.getWeatherResult(cities)

    expect(mockAxios.get).toHaveBeenCalledTimes(3)
    expect(mockAxios.get).toHaveBeenCalledWith('/weather', {
      params: {
        q: 'Cairo'
      }
    })
    expect(mockAxios.get).toHaveBeenCalledWith('/weather', {
      params: {
        zip: '10005'
      }
    })
    expect(mockAxios.get).toHaveBeenCalledWith('/weather', {
      params: {
        q: 'New York'
      }
    })

    expect(results).toMatchSnapshot()
  })
})
