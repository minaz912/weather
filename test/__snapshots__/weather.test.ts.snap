// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getWeatherResult returns weather result messages based on given cities 1`] = `
Array [
  Object {
    "name": "London",
    "timezoneOffsetInSeconds": "7200",
    "weather": "Drizzle, 280.32 Celsius",
  },
  Object {
    "name": "London",
    "timezoneOffsetInSeconds": "7200",
    "weather": "Drizzle, 280.32 Celsius",
  },
  Object {
    "name": "London",
    "timezoneOffsetInSeconds": "7200",
    "weather": "Drizzle, 280.32 Celsius",
  },
]
`;
